src/Transports.AspNetCore/GraphQLHttpMiddleware.cs

```diff
@@ -47,31 +41,36 @@ public async Task InvokeAsync(HttpContext context)
 
             // Handle requests as per recommendation at http://graphql.org/learn/serving-over-http/
             var httpRequest = context.Request;
-            var gqlRequest = new GraphQLRequest();
+            GraphQLRequest gqlRequest = null;
             GraphQLRequest[] gqlBatchRequest = null;
 
             var writer = context.RequestServices.GetRequiredService<IDocumentWriter>();
+            var cancellationToken = GetCancellationToken(context);
 
-            if (HttpMethods.IsGet(httpRequest.Method) || (HttpMethods.IsPost(httpRequest.Method) && httpRequest.Query.ContainsKey(GraphQLRequest.QueryKey)))
+            if (HttpMethods.IsGet(httpRequest.Method) ||
+                (HttpMethods.IsPost(httpRequest.Method) && httpRequest.Query.ContainsKey(GraphQLRequest.QueryKey)))
             {
-                ExtractGraphQLRequestFromQueryString(httpRequest.Query, gqlRequest);
+                gqlRequest = _deserializer.DeserializeFromQueryString(httpRequest.Query);
             }
             else if (HttpMethods.IsPost(httpRequest.Method))
             {
                 if (!MediaTypeHeaderValue.TryParse(httpRequest.ContentType, out var mediaTypeHeader))
                 {
-                    await WriteBadRequestResponseAsync(context, writer, $"Invalid 'Content-Type' header: value '{httpRequest.ContentType}' could not be parsed.").ConfigureAwait(false);
+                    await WriteBadRequestResponseAsync(context, writer, cancellationToken, $"Invalid 'Content-Type' header: value '{httpRequest.ContentType}' could not be parsed.").ConfigureAwait(false);
                     return;
                 }
 
                 switch (mediaTypeHeader.MediaType)
                 {
                     case JsonContentType:
-                        if (!Deserialize(httpRequest.Body, out gqlRequest, out gqlBatchRequest))
+                        var deserializationResult = await _deserializer.DeserializeFromJsonBodyAsync(httpRequest, cancellationToken).ConfigureAwait(false);
+                        if (!deserializationResult.IsSuccessful)
                         {
-                            await WriteBadRequestResponseAsync(context, writer, "Body text could not be parsed. Body text should start with '{' for normal graphql query or with '[' for batched query.").ConfigureAwait(false);
+                            await WriteBadRequestResponseAsync(context, writer, cancellationToken, "Body text could not be parsed. Body text should start with '{' for normal graphql query or with '[' for batched query.").ConfigureAwait(false);
                             return;
                         }
+                        gqlRequest = deserializationResult.Single;
+                        gqlBatchRequest = deserializationResult.Batch;
                         break;
 
                     case GraphQLContentType:
```

Comment:
Please add test for this `ContentType=application/graphql` . I see that dereferencing `gqlRequest.Query` now gives NRE.

Reply:
Yep, good spot. Problem is we can't just `new` up a `GraphQLRequest` in the middleware anymore since it only sees the abstract class. Will need to have the `IGraphQLRequestDeserializer` interface implement one, and can probably do it as a default interface method I guess, unless that sounds horrible to you? :P

Reply:
Can't do it as a default interface implementation as it's still just an abstract type there, would need an interface method of `CreateDefault()` like I had back last year. Will just have to live with a duped streamreader line of code in each project. Not really a biggy.

Unless I made GraphQLRequest a normal class,... hmm

Reply:
Have fixed this and there's a test.

Reply:
You do not need full `GraphQLRequest` in this case, all other properties are `null`.

Reply:
>  can probably do it as a default interface method I guess, unless that sounds horrible to you?

for me it sounds like using the wrong tool to solve a problem